src/cache_cf.cc

```diff
@@ -1018,87 +1036,135 @@ parse_obsolete(const char *name)
+        throw TexcHere(ToSBuf("the time must be less than ", maxYears, " years"));
     }
 }
 
-static uint64_t
-parseTimeUnits(const char *unit, bool allowMsec)
+template <class TimeUnit>
+static TimeUnit
+FromNanoseconds(const std::chrono::nanoseconds &ns, const double parsedValue)
 {
-    if (allowMsec && !strncasecmp(unit, T_MILLISECOND_STR, strlen(T_MILLISECOND_STR)))
-        return 1;
+    const auto result = std::chrono::duration_cast<TimeUnit>(ns);
+    if (!result.count()) {
+        throw TexcHere(ToSBuf("time value '", parsedValue,
+                    "' is too small to be used in this context, the minimal value is 1 ",
+                    TimeUnitToString<TimeUnit>()));
+    }
+    return result;
+}
 
-    if (!strncasecmp(unit, T_SECOND_STR, strlen(T_SECOND_STR)))
-        return 1000;
+/// Parses a time specification from the config file and
+/// returns the time as a chrono duration object of 'TimeUnit' type.
+template <class TimeUnit>
+static TimeUnit
+parseTimeLine()
+{
+    const auto valueToken = ConfigParser::NextToken();
+    if (!valueToken) {
+        self_destruct();
+        return TimeUnit::zero();
+    }
 
-    if (!strncasecmp(unit, T_MINUTE_STR, strlen(T_MINUTE_STR)))
-        return 60 * 1000;
+    const auto parsedValue = xatof(valueToken);
 
-    if (!strncasecmp(unit, T_HOUR_STR, strlen(T_HOUR_STR)))
-        return 3600 * 1000;
+    if (parsedValue == 0)
+        return TimeUnit::zero();
 
-    if (!strncasecmp(unit, T_DAY_STR, strlen(T_DAY_STR)))
-        return 86400 * 1000;
+    std::chrono::nanoseconds parsedUnitDuration;
 
-    if (!strncasecmp(unit, T_WEEK_STR, strlen(T_WEEK_STR)))
-        return 86400 * 7 * 1000;
+    const auto token = ConfigParser::PeekAtToken();
 
-    if (!strncasecmp(unit, T_FORTNIGHT_STR, strlen(T_FORTNIGHT_STR)))
-        return 86400 * 14 * 1000;
+    if (!parseTimeUnit<TimeUnit>(token, parsedUnitDuration))
+        throw TexcHere(ToSBuf("unknown time unit '", token, "'"));
 
-    if (!strncasecmp(unit, T_MONTH_STR, strlen(T_MONTH_STR)))
-        return static_cast<uint64_t>(86400) * 30 * 1000;
+    (void)ConfigParser::NextToken();
 
-    if (!strncasecmp(unit, T_YEAR_STR, strlen(T_YEAR_STR)))
-        return static_cast<uint64_t>(86400 * 1000 * 365.2522);
+    CheckTimeValue(parsedValue, parsedUnitDuration);
 
-    if (!strncasecmp(unit, T_DECADE_STR, strlen(T_DECADE_STR)))
-        return static_cast<uint64_t>(86400 * 1000 * 365.2522 * 10);
+    const std::chrono::nanoseconds nanoseconds(static_cast<std::chrono::nanoseconds::rep>(parsedUnitDuration.count() * parsedValue));
 
-    debugs(3, DBG_IMPORTANT, "parseTimeUnits: unknown time unit '" << unit << "'");
+    // validate precisions (time-units-small only)
+    if (TimeUnit(1) <= std::chrono::microseconds(1)) {
+        if (0 < nanoseconds.count() && nanoseconds.count() < 3) {
+            debugs(3, DBG_CRITICAL, "WARNING: " << ConfigPosition() <<
```

Comment:
Since this is a WARNING rather than an ERROR, perhaps it should be DBG_IMPORTANT? Or is there a different set of rules for parsing-related error levels? Also, please check whether we are supposed to use DBG_PARSE_NOTE() here and in similar new code.

```suggestion
            debugs(3, DBG_IMPORTANT, ConfigPosition << "WARNING: " <<
```

Reply:
> DBG_IMPORTANT does not work during configuration parsing

If you agree that `DBG_IMPORTANT` is more appropriate for warnings, then I suggest using `DBG_PARSE_NOTE(DBG_IMPORTANT)` here and filing a bug report that `DBG_PARSE_NOTE(DBG_IMPORTANT)` does not work during initial configuration parsing.

It is much easier to fix the bug than to find all the places where folks abused `DBG_CRITICAL` to work around that bug.

Reply:
Ok, I created a bug [report](https://bugs.squid-cache.org/show_bug.cgi?id=4971).